README
This is aethertool, a command-line tool for aether websites
Copyright (C) 2005 Jeff Epler <jepler@unpythonic.net>

REQUIREMENTS
    Python 2.3 or above
                to execute the scripts

    convert     to resize images

    pngcrush    to improve the compression of png images

    jpegtran    to improve the compression of jpeg images and perform lossless
                rotation of digital camera pictures when EXIF orientation data
                is available


INSTALLATION

    To do a system-wide installation:
        $ sudo python setup.py install    

    To do a single-user installation:
        $ python setup.py install --home ~
    You must have ~/bin on $PATH, and ~/lib/python on $PYTHONPATH

    For more installation options, use
        $ python setup.py --help
    or refer to Installing Python Modules, http://docs.python.org/dist/

    Aethertool can also be used "in place" without installation.
    

USAGE

    For help with commandline options, use "aethertool.py -h".  For help on the
    configuration syntax, use "aethertool.py -c help".


BUGS

    Aethertool assumes that your browser is called "firefox"; it should use
    the webbrowser module instead

    Aethertool assumes the command used to invoke the editor doesn't exit
    until the user is done editing the file.  This can cause problems if the
    editor naturally starts in the background.

    If the user does a 'write and exit' sequence, the file the browser loads
    to submit page changes may be removed too early, and the last edit lost.


COPYRIGHT

    This program is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the Free
    Software Foundation; either version 2 of the License, or (at your option)
    any later version.

    This program is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
    FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
    more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc., 59
    Temple Place, Suite 330, Boston, MA  02111-1307  USA


debian/changelog
aethertool (0.9.5+nmu1) stable; urgency=medium

  * Remove debug print

 -- jepler <jepler@babs.unpythonic.net>  Wed, 11 Nov 2015 21:45:51 -0600

aethertool (0.9.5) stable; urgency=low

  * Fix file uploads with https

 -- Jeff Epler <jepler@dsndata.com>  Wed, 11 Nov 2015 08:15:16 -0600

aethertool (0.9.4) testing; urgency=medium

  * Debian jessie moved my cheese

 -- jepler <jepler@babs.unpythonic.net>  Wed, 11 Mar 2015 21:19:58 -0500

aethertool (0.9.3) stable; urgency=medium

  * Fix -n bug

 -- jepler <jepler@babs.unpythonic.net>  Sat, 22 Feb 2014 21:22:02 -0600

aethertool (0.9.2) stable; urgency=medium

  * Improvements

 -- jepler <jepler@babs.unpythonic.net>  Sat, 22 Feb 2014 21:18:21 -0600

aethertool (0.9.1) squeeze; urgency=low

  * Upgrade from ClientForm to mechanize

 -- jepler <jepler@platte.unpythonic.net>  Sat, 07 Apr 2012 20:45:24 -0500

aethertool (0.9) squeeze; urgency=low

  * cooperate with python-clientform package

 -- jepler <jepler@unpythonic.net>  Mon, 15 Mar 2010 17:48:01 -0500

aethertool (0.8) karmic; urgency=low

  * fix problem with special chars in passwords

 -- jepler <jepler@unpythonic.net>  Mon, 15 Mar 2010 17:48:01 -0500

aethertool (0.7) unstable; urgency=low

  * Make suffix for blog entry names optional
  * Fixes for non-ASCII page contents
  * Include clientform in the distribution
  * New '-U' mode prints edit URL
  * Produce medium (default: 900x900) reduced images
  * Don't produce thumbs or medium images if the original is smaller (in pixels)
  * Produce thumbs or medium images as jpg if is smaller than original type
    (in bytes)
  * Allow URLs without index.cgi
  * Speed up adding attachments to pages with many attachments
  * Allow specification of site by URL instead of site identifier

 -- Jeff Epler <jepler@unpythonic.net>  Tue, 08 Sep 2009 21:15:11 -0500

aethertool (0.6) unstable; urgency=low

  * Initial release.

 -- Jeff Epler <jepler@unpythonic.net>  Mon, 26 May 2008 21:49:30 -0500


mkdist.py
#!/usr/bin/python
# vim: set fileencoding=utf-8 : sts=4 : et : sw=4
#    This is a component of mailpie, a full-text search for email
#
#    Copyright © 2008 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import commands
import gzip
import os
import re
import sys
import tarfile
import time
import StringIO

version_tag = re.compile("^v\d+(\.\d+)*$")

def tag_parse(t):
    return [int(ti) for ti in t[1:].split(".")]

def tag_pad(s, l):
    return s + [0] * (l - len(s))

def tag_sort(a, b):
    a = tag_parse(a)
    b = tag_parse(b)

    a = tag_pad(a, len(b))
    b = tag_pad(b, len(a))

    return cmp(a, b)

def highest_version():
    tags = os.popen("git tag").readlines()
    tags = [t.strip() for t in tags if version_tag.match(t)]
    tags.sort(cmp=tag_sort)
    return tags[-1]

def main(args):
    if len(args) == 1:
        version = args[0]
    elif len(args) > 1:
        raise SystemExit("Usage: %s [version]" % sys.argv[0])
    else:
        status, gitversion = commands.getstatusoutput("git-describe --tags")
        version = highest_version()
        if status != 0 or version != gitversion:
            raise SystemExit("""\
Highest version %r doesn't match description %r.
Specify version number explicitly if this is what you want""" % (
                    version, gitversion))

    version = version.lstrip("v")

    DIRNAME = "%(p)s-%(v)s" % {'p': 'mailpie', 'v': version}
    TARNAME = DIRNAME + '.tar.gz'

    verstream = StringIO.StringIO("%s\n" % version)
    verinfo = tarfile.TarInfo(DIRNAME + "/VERSION")
    verinfo.mode = 0660
    verinfo.size = len(verstream.getvalue())
    verinfo.mtime = time.time()

    tardata = os.popen("git-archive --prefix=%(p)s/ v%(v)s"
                            % {'p': DIRNAME, 'v': version}).read()
    tarstream = StringIO.StringIO(tardata)

    tar = tarfile.TarFile(mode="a", fileobj=tarstream)
    tar.addfile(verinfo, verstream)
    tar.close()

    out = gzip.open("../" + TARNAME, "wb")
    out.write(tarstream.getvalue())
    out.close()
    os.system("ls -l ../%s" % TARNAME)

if __name__ == '__main__':
    main(sys.argv[1:])


disorient.py
#   This is a component of aethertool, a command-line interface to aether
#   Copyright 2005-2021 Jeff Epler <jepler@unpythonic.net>
#
#   This program is free software: you can redistribute it and/or modify it
#   under the terms of the GNU General Public License version 3 as published by
#   the Free Software Foundation.
#
#   This program is distributed in the hope that it will be useful, but WITHOUT
#   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
#   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
#   more details.
#
#   You should have received a copy of the GNU General Public License along
#   with this program.  If not, see <http://www.gnu.org/licenses/>.

# This program is a new implementation of the method described at
#  http://sylvana.net/jpegcrop/exif_orientation.html
# and in the "jpegexiforient.c" source code.

EXIF_MAGIC = "\xff\xd8\xff\xe1"  #  JPEG SOI + Exif APP1
EXIF_MAGIC_2 = "Exif\0\0"

import struct


def read_2(f):
    return ord(f.read(1)) * 256 + ord(f.read(1))


def exif_orientation(f, set=None):
    if set:
        f = open(f, "rb+")
    else:
        f = open(f, "rb")

    buf = f.read(4)
    # print `buf`
    if buf != EXIF_MAGIC:
        return

    length = read_2(f)
    # print `length`
    if length <= 20:
        return

    buf = f.read(6)
    # print `buf`
    if buf != EXIF_MAGIC_2:
        return

    buf = f.read(length - 8)
    # print `buf[:2]`
    if buf[:2] == "II":
        is_motorola, order = 0, "<"
    elif buf[:2] == "MM":
        is_motorola, order = 1, ">"
    else:
        return

    decode_4 = lambda i: struct.unpack(order + "I", buf[i : i + 4])[0]
    decode_2 = lambda i: struct.unpack(order + "H", buf[i : i + 2])[0]
    # print "here", repr(buf[2:4])
    if is_motorola:
        if buf[2:4] != "\0\x2a":
            return
    else:
        if buf[2:4] != "\x2a\0":
            return

    offset = decode_4(4)
    # print offset
    if offset > length - 2:
        return

    number_of_tags = decode_2(offset)
    offset += 2
    # print "here", offset, length

    while 1:
        if offset > length - 12:
            return
        tagnum = decode_2(offset)
        # print "loop", offset, length, tagnum
        if tagnum == 0x112:
            break
        number_of_tags -= 1
        if not number_of_tags:
            return 0
        offset += 12

    if not set:
        # print repr(buf[offset:offset+12])
        return decode_2(offset + 8)
    else:
        if is_motorola:
            new_tag = "\0\3" "\0\0\0\1" "\0" + chr(set) + "\0\0"
        else:
            new_tag = "\3\0" "\1\0\0\0" + chr(set) + "\0" "\0\0"
    f.seek(4 + 2 + 6 + 2 + offset)
    f.write(new_tag)
    f.seek(0, 2)
    # print "replaced"


def clear_exif_orientation(x):
    return exif_orientation(x, 1)


descr = {
    1: "Normal",
    2: "Mirrored left-to-right",
    3: "Rotated 180 degrees",
    4: "Mirrored top-to-bottom",
    5: "Mirrored along top-left diagonal",
    6: "Rotated 90 degrees",
    7: "Mirrored along top-right diagonal",
    8: "Rotated 270 degrees",
}

if __name__ == "__main__":
    import sys

    if sys.argv[1] == "-c":
        clear_exif_orientation(sys.argv[2])
        print("Orientation cleared")
    else:
        print(descr.get(exif_orientation(sys.argv[1]), "Invalid orientation"))
# vim:sw=4:sts=4:et


aethertool.py
#!/usr/bin/env python
#    This is a component of aethertool, a command-line interface to aether
#    Copyright 2005 Jeff Epler <jepler@unpythonic.net>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import getpass, urllib2, ClientForm, tempfile, os, time, webbrowser, sys
import getopt, cgi, re, disorient
import shutil, atexit
from sys import argv
from htmlentitydefs import name2codepoint

EDITOR = os.environ.get("EDITOR", "vim")

thumb_geometry = "300x300"
medium_geometry = "900x900"

config = {}
default_config = 'default'
browser_wait = 5
def add_config(name, root, password=None, thumb_geometry=None):
    config[name] = (root, password, thumb_geometry)

def set_default(name):
    global default_config
    default_config = name

def load_config(config):
    global AETHER_TOP, PASS, thumb_geometry
    AETHER_TOP = config[0]
    PASS = config[1] or getpass.getpass()
    thumb_geometry = config[2] or thumb_geometry

rcfile = os.path.join(os.environ.get('HOME', ''), ".aetherrc")
if os.path.exists(rcfile): execfile(rcfile)
else:
    print ("The configuration file %r does not exist.\n"
           "aethertool cannot run without it." % rcfile)
    raise SystemExit

config_name = os.path.splitext(os.path.split(sys.argv[0])[1])[0]
if len(config) == 1: default_config = config.keys()[0]
if not config.has_key(config_name): config_name = default_config

def quote_paranoid(text):
    """ Convert utf-8 string to sequence of lower case English characters. """

    text = text.encode('utf-8')

    result = ''
    for char in text:
        result += chr(ord('a') + ord(char)/16) +\
                  chr(ord('a') + ord(char)%16)

    return result
import httplib, mimetypes

def post_multipart(host, selector, fields, files):
    """
    Post fields and files to an http host as multipart/form-data.
    fields is a sequence of (name, value) elements for regular form fields.
    files is a sequence of (name, filename, value) elements for data to be uploaded as files
    Return the server's response page.
    """
    content_type, body = encode_multipart_formdata(fields, files)
    h = httplib.HTTPConnection(host)
    headers = {
        'Content-Type': content_type,
        'Content-Length': len(body),
        'User-Agent': 'Aethertool/0.9'}
    h.request('POST', selector, body, headers)
    #h.putrequest('POST', selector)
    #h.putheader('content-type', content_type)
    #h.putheader('content-length', str(len(body)))
    #h.endheaders()
    #h.send(body)
    res = h.getresponse()
    #errcode, errmsg, headers = h.getreply()
    if res.status != 200:
        raise RuntimeError, (res.status, res.reason, res.read())
    return res.read()

def encode_multipart_formdata(fields, files):
    """
    fields is a sequence of (name, value) elements for regular form fields.
    files is a sequence of (name, filename, value) elements for data to be uploaded as files
    Return (content_type, body) ready for httplib.HTTP instance
    """
    BOUNDARY = '----------ThIs_Is_tHe_bouNdaRY_$'
    CRLF = '\r\n'
    L = []
    for (key, value) in fields:
        L.append('--' + BOUNDARY)
        L.append('Content-Disposition: form-data; name="%s"' % key)
        L.append('')
        L.append(value)
    for (key, filename, value) in files:
        L.append('--' + BOUNDARY)
        L.append('Content-Disposition: form-data; name="%s"; filename="%s"' % (key, filename))
        L.append('Content-Type: %s' % get_content_type(filename))
        L.append('')
        L.append(value)
    L.append('--' + BOUNDARY + '--')
    L.append('')
    body = CRLF.join(L)
    content_type = 'multipart/form-data; boundary=%s' % BOUNDARY
    return content_type, body

def get_content_type(filename):
    return mimetypes.guess_type(filename)[0] or 'application/octet-stream'

import urlparse

def posturl(url, fields, files):
    urlparts = urlparse.urlsplit(url)
    return post_multipart(urlparts[1], urlparts[2], fields,files)

def webbrowser_open(u):
    os.spawnvp(os.P_NOWAIT, "firefox", ["firefox", u])
    time.sleep(browser_wait)

def get_edit_url(page):
    return AETHER_TOP + "?action=edit&password=%s&name=%s" % (PASS, page)

def get_edit_form(page):
    url = get_edit_url(page)

    forms = ClientForm.ParseResponse(urllib2.urlopen(url))
    for f in forms:
        try:
            action = f.get_value("action")
        except ClientForm.ControlNotFoundError:
            continue
        if action == "edit":
            return f
    print "Required form not found.  AETHER_TOP or password may be incorrect."

entity_re = re.compile("&(#[0-9]+|[a-zA-Z0-9]+);?")

def unescape_replace(m):
    g = m.group(1)
    if g.startswith("#"):
        return unichr(int(g[1:])).encode('utf-8')
    codepoint = name2codepoint.get(g)
    if codepoint is None: return g
    return unichr(codepoint).encode('utf-8')

def unescape(s):
    return entity_re.sub(unescape_replace, s)

current_text_pat = re.compile(
    "<textarea[^>]*>"
    "([^<]*)"
    "</textarea>", re.I | re.M | re.DOTALL)

def get_current_text(page):
    url = get_edit_url(page)
    page = urllib2.urlopen(url).read()
    m = current_text_pat.search(page).group(1)
    m = unescape(m)
    return m

def save_text(page, newtext):
    f = get_edit_form(page)
    f['text'] = newtext
    return f.click("save")

def preview_text(page, newtext):
    return """<HTML>
    <META http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <BODY onload="document.forms[0].submit()"
    <DIV style="display: none">
    <FORM method=post accept-charset=\"UTF-8\" action=\"%s#buttonsandpreview\">
        <input type=hidden name=action value=edit>
        <input type=hidden name=password value=\"%s\">
        <input name=newname value=\"%s\">
        <input name=name value=\"%s\">
        <input name=hasnewname value=yes>
        <textarea name=text>%s</textarea>
        <input type=submit name=preview>
    </form>
    </DIV>
    Submitting updated text...
    </body>""" % (
    cgi.escape(AETHER_TOP, True),
    cgi.escape(PASS, True),
    cgi.escape(page, True),
    cgi.escape(page, True),
    cgi.escape(newtext, True))

def upload_file(page, local, remote, quiet=0):
    qPASS = quote_paranoid(PASS)
    qpage = quote_paranoid(page)
    url = AETHER_TOP
    data = [('action', 'attach'), ('password', qPASS), ('name', qpage), ('nolisting', '1')]
    fdata = [('file', remote, open(local, "rb").read())]
    posturl(url, data, fdata)
    if not quiet:
        print "Uploaded file is:"
        print "   [page %s] [file %s]" % (page, remote)
        if AETHER_TOP.endswith("/"):
            print "   " + AETHER_TOP + "files/%s/%s" % (page, remote)
        else:
            print "   " + AETHER_TOP + "-files/%s/%s" % (page, remote)

def put_page(page, filename):
    if hasattr(filename, 'read'):
        contents = filename.read()
    elif filename is None:
        contents = ''
    else:
        contents = open(filename).read()
    print "size of new contents", len(contents)
    u = save_text(page, contents)
    try:
        urllib2.urlopen(u)
    except urllib2.HTTPError, detail:
        if detail.code != 404 or contents:
            raise

def edit_page(page):
    open(os.path.expanduser("~/.aetherlast"), "w").write(page)

    t = get_current_text(page)

    name = os.path.join(tempdir, (page.replace("/", "_") or "index") + ".ae")
    fd = open(name, "w")
    fd.write(t)
    fd.close()

    hname = os.path.join(tempdir, "submit.html")

    pid = os.spawnvp(os.P_NOWAIT, EDITOR, [EDITOR, name])

    ost = os.stat(name)
    while 1:
        nst = os.stat(name)
        if ost.st_mtime != nst.st_mtime:
            ost = nst
            new_text = open(name).read()
            u = preview_text(page, new_text)
            hfd = open(hname, "w")
            hfd.write(u)
            hfd.close()
            webbrowser_open(hname) 
        sts = os.waitpid(0, os.P_NOWAIT)
        if sts[0] == pid:
            break
        time.sleep(1)
    os.unlink(name)

def isimage(f):
    ext = os.path.splitext(f)[1].lower()
    if ext in ['.jpg', '.jpeg', '.png', '.gif']:
        return True
    return False

def get_orient(f):
    return disorient.exif_orientation(f) or 1

def optimize(f):
    ext = os.path.splitext(f)[1].lower()
    t = os.path.join(tempdir, "optimized" + ext)
    if ext in ['.jpg', '.jpeg']:
        command = ['jpegtran', '-optimize', '-progressive',
            '-copy', 'all', '-outfile', t]
        o = get_orient(f)
        if o == 8:
            command.extend(['-rotate', '270'])
        elif o == 6:
            command.extend(['-rotate', '90'])
        command.append(f)
        os.spawnvp(os.P_WAIT, command[0], command)
        if o in (6, 8): disorient.clear_exif_orientation(t)
        return t
    if ext in ['.png']:
        os.spawnvp(os.P_WAIT, "pngcrush", ["pngcrush", "-q", "-reduce", f, t])
        return t
    # XXX convert gif to png?
    return f

tempdir = tempfile.mkdtemp()
atexit.register(shutil.rmtree, tempdir)

def usage(exitcode=1):
    print """\
Usage:
    %(name)s [-e] page
        Edit 'page'.  Use '/' (forward slash) to edit the front page.  To
        preview the changes in your browser, save the file.  Use the browser's
        save button in the browser to save the changes.
    %(name)s -p page < contents
        Put new contents to 'page' from standard input.
    %(name)s -n [-k suffix] blog
        Create a new blog entry on 'blog', optonally with 'suffix' added
        to the name of the created page.
    %(name)s -u [-t] [-g WxH] page file[=localfile] file...
    %(name)s -u [-t] [-g WxH] -l file[=localfile] file...
        Upload files to 'page' or the last page edited (-l).  Create thumbnails
        unless -t is specified.  Use -g to specify the maximum thumbnail size,
        currently %(thumbsize)s.
    %(name)s -d page
        Delete 'page'
    %(name)s -c configname [other usage from above]
        Use the configuration 'configname' instead of the default
        For help on the syntax of configuration files, use "-c help".
""" % {'name': os.path.basename(sys.argv[0]), 'thumbsize': thumb_geometry}
    raise SystemExit, exitcode

def help_config():
    print """\
The configuration file ~/.aetherrc is a python script.  A typical one might 
look like this (without the indentation):
    add_config('configname', 'http://www.example.com/index.cgi', 'password')
    set_default('configname')
The configuration is guessed from the script's name, otherwise the value
given by set_default is used.  If only one call to add_configuration is
present, then that configuration is the default.

If 'password' is not specified, then it is prompted when the script is run.
"""
    print "The following configurations are defined:"
    for k in config: print "\t" + k
    print
    print "When invoked as '%s', the default configuration is '%s'" % (
        os.path.basename(sys.argv[0]), default_config)
    raise SystemExit, 0

try:
    opts, args = getopt.getopt(argv[1:], "+c: denpul U t m:k: g: c: h?")
except getopt.GetoptError, detail:
    print >> sys.stderr, "%s:" % os.path.basename(sys.argv[0]), detail
    usage()

MODE_EDIT, MODE_NEW_ENTRY, MODE_PUT, MODE_UPLOAD, MODE_DELETE, MODE_URL = range(6)

mode = MODE_EDIT
suffix = ""
do_thumbnail = True

for k, v in opts:
    if k == "-c": config_name = v

    if k == "-d": mode = MODE_DELETE
    if k == "-e": mode = MODE_EDIT
    if k == "-n": mode = MODE_NEW_ENTRY
    if k == "-p": mode = MODE_PUT
    if k == "-u": mode = MODE_UPLOAD

    if k == "-U": mode = MODE_URL

    if k == "-t": do_thumbnail = not do_thumbnail
    if k == "-k": suffix = "-" + v.replace(" ", "-")
    if k == "-g": thumb_geometry = v
    if k == "-m": medium_geometry = v
    if k == "-?" or k == "-h": usage(0)

    if k == "-l": 
        args.insert(0, open(os.path.expanduser("~/.aetherlast")).read().strip())
if config_name == "help":
    help_config()

load_config(config[config_name])

if mode == MODE_URL:
    print get_edit_url("")
elif mode == MODE_UPLOAD:
    page = args[0]
    for filename in args[1:]:
        if "=" in filename:
            remote, local = filename.split("=", 1)
        else:
            remote = os.path.basename(filename)
	    if not re.search('[a-z]', remote): remote = remote.lower()
            local = filename
	if isimage(remote):
	    local = optimize(local)
        upload_file(page, local, remote)
        if do_thumbnail and isimage(remote):
            base, ext = os.path.splitext(remote)

            med = base + "-medium" + ext
            localmed = os.path.join(tempdir, "medium" + ext)
            #print "Creating medium image for", remote
            os.spawnvp(os.P_WAIT, "convert", 
                ["convert", "-geometry", medium_geometry, local, localmed])
            localmed = optimize(localmed)
            upload_file(page, localmed, med)

            thumb = base + "-small" + ext
            localthumb = os.path.join(tempdir, "thumb" + ext)
            #print "Creating thumbnail for", remote
            os.spawnvp(os.P_WAIT, "convert", 
                ["convert", "-geometry", thumb_geometry, local, localthumb])
            localthumb = optimize(localthumb)
            upload_file(page, localthumb, thumb)

elif mode == MODE_NEW_ENTRY:
    if args:
        page = args[0] + "/0" + str(int(time.time())) + suffix
    else:
        page = "0" + str(int(time.time())) + suffix
    edit_page(page)
elif mode == MODE_EDIT:
    if args:
        page = args[0]
    else:
        page = "sandbox"
    page = page.strip("/")
    edit_page(page)
elif mode == MODE_PUT:
    page = args[0].strip("/")
    put_page(page, sys.stdin)
elif mode == MODE_DELETE:
    page = args[0].strip("/")
    put_page(page, None)
    
# vim:sw=4:sts=4:et
